core/lexer.h

```diff
@@ -113,13 +117,76 @@ static inline std::ostream &operator<<(std::ostream &o, const FodderElement &f)
 
 /** A sequence of fodder elements, typically between two tokens.
  *
- * A LINE_END is not allowed to follow a PARAGRAPH or a LINE_END.  This can be represented by
+ * A LINE_END is not allowed to follow a PARAGRAPH or a LINE_END. This can be represented by
  * replacing the indent of the prior fodder and increasing the number of blank lines if necessary.
  * If there was a comment, it can be represented by changing the LINE_END to a paragraph containing
  * the same single comment string.
+ *
+ * There must be a LINE_END or a PARAGRAPH before a PARAGRAPH.
```

Comment:
I think that actually this is a good thing after all.

Some reasons starting paragraphs on a clean line is a good thing (I think you've mentioned most of them in our discussion):
* Having paragraphs starting mid-line is a weird input and boundary case for all the passes that touch fodders and newlines. It's more "hygienic" to eliminate these cases completely.
* We don't allow paragraphs to end mid-line (which is a very good thing, as it would lead to even weirder cases and it would prevent clean conversion to single-line comments). So symmetry here is nice.
* I have randomly asked some people and apparently no one can think of a reason why someone would want to have such thing in their code.

Some reasons why the current representation is reasonable (compared to e.g. treating the line before as part of the paragraph):
* How to otherwise represent the line comment that can go in the line before? Keeping it in a paragraph is very ugly (it effectively gets rid of the invariant + allows mixing comment styles within a paragraph). Some rule like "PARAGRAPH starts with an endline unless there is LINE_END before it" is better, but still a little confusing.
* Why not just another pass? Because having these newlines affects other passes, for example FixIndentation and SortImports . So we want to keep this as invariant at all times and not just when we're done. 
* Some existing code depends on it. There's a good chance some bugs would be introduced while changing it. I'm not really convinced that the alternatives I can think of now are actually better so keeping things as they are seems like a reasonable default. Especially since this PR is about something else.